skipchain/skipchain.go

```diff
@@ -418,47 +423,53 @@ func (s *Service) bftVerifyFollowBlock(msg []byte, data []byte) bool {
 // is valid.
 func (s *Service) bftVerifyNewBlock(msg []byte, data []byte) bool {
 	log.Lvlf4("%s verifying block %x", s.ServerIdentity(), msg)
-	src := data[0:32]
-	srcSB := s.Sbm.GetByID(src)
-	if srcSB == nil {
+	srcHash := data[0:32]
+	prevSB := s.Sbm.GetByID(srcHash)
+	if prevSB == nil {
 		log.Error("Didn't find src-skipblock")
 		return false
 	}
-	_, dstN, err := network.Unmarshal(data[32:])
+	_, newSBi, err := network.Unmarshal(data[32:])
 	if err != nil {
 		log.Error("Couldn't unmarshal SkipBlock", data)
 		return false
 	}
-	dst := dstN.(*SkipBlock)
-	if !dst.Hash.Equal(SkipBlockID(msg)) {
-		log.Lvlf2("Dest skipBlock different from msg %x %x", msg, []byte(dst.Hash))
+	newSB := newSBi.(*SkipBlock)
+	if !newSB.Hash.Equal(SkipBlockID(msg)) {
+		log.Lvlf2("Dest skipBlock different from msg %x %x", msg, []byte(newSB.Hash))
 		return false
 	}
 
-	isFree := func() bool {
-		for i, b := range dst.BackLinkIDs {
-			if b.Equal(src) {
-				return srcSB.GetForwardLen() == i
-			}
-		}
+	if !newSB.BackLinkIDs[0].Equal(srcHash) {
+		log.Lvl2("Backlink does not point to previous block:", prevSB.Index, newSB.Index)
 		return false
-	}()
-	if !isFree {
-		log.Lvl2("Didn't find a free corresponding forward-link")
+	}
+	if len(prevSB.ForwardLink) > 0 {
+		log.Lvl2("previous block already has forward-link")
+		return false
+	}
+	if !s.newBlocks.acceptSign(prevSB.Hash, newSB.Hash) {
+		log.Lvl2("previous block points to another block")
 		return false
 	}
 
-	for _, ver := range dst.VerifierIDs {
-		f, ok := s.verifiers[ver]
-		if !ok {
-			log.Lvlf2("Found no user verification for %x", ver)
-			return false
-		}
-		if !f(msg, dst) {
-			return false
+	ok := func() bool {
+		for _, ver := range newSB.VerifierIDs {
+			f, ok := s.verifiers[ver]
+			if !ok {
+				log.Lvlf2("Found no user verification for %x", ver)
+				return false
+			}
+			if !f(msg, newSB) {
+				return false
+			}
 		}
+		return true
+	}()
```

Comment:
You like thoses anonymous functions :D  But here it can be made a bit simpler still in my opinion.
```go
ok := true 
 for _, ver := range newSB.VerifierIDs {
        f, ok := s.verifiers[ver]
 	if !ok || !f(msg, newSB) {
                ok = false
                break
 	}
  }
```